pkg/renderer/sgml/renderer.go

```diff
@@ -0,0 +1,360 @@
+package sgml
+
+import (
+	"bytes"
+	htmltemplate "html/template"
+	"io"
+	"strings"
+	texttemplate "text/template"
+
+	"github.com/bytesparadise/libasciidoc/pkg/configuration"
+	"github.com/bytesparadise/libasciidoc/pkg/types"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// Renderer implements the backend render interface by using sgml.
+type Renderer interface {
+
+	// Render renders a document to the given output stream.
+	Render(ctx *Context, doc Document, output io.Writer) (Metadata, error)
+
+	// SetFunction sets the named function.
+	SetFunction(name string, fn interface{})
+
+	// Templates returns the Templates used by this Renderer.
+	// It cannot be altered on a given Renderer, since the old
+	// templates may have already been parsed.
+	Templates() Templates
+}
+
+func NewRenderer(t *Templates) Renderer {
+	sr := &sgmlRenderer{
+		functions: make(funcMap),
+		templates: t, // TODO: should we copy this?
```

Comment:
instead of copying here, you could avoid passing it as a pointer above.
By the way, I've tried passing as values rather than pointers as much as possible (I believe it's more efficient), unless it's absolutely needed (ie, the argument is going to be changed by the func).

Reply:
So that's actually a fallacy, and IMO misguided.

Copy by value beats indirection in cases where you either are going to dereference every member anyway (so you're already copying anyway), or the object in question is small.

However, the total time difference is probably on the order of a few nanoseconds, and unless the dereferencing is done multiple times per member (it isn't here), it's likely a net wash (not even a few ns).

(For the record, I've spent a significant part of my life building very high performance software for use in circumstances like high frequency trading, IP protocol routing, etc.  At one point I was employed by Sun Microsystems as a contractor to optimize network performance to help them reach higher packet-per-second counts for IP forwarding.  A significant part of the effort was often "undoing" attempts to optimize code paths that ultimately lead to more complexity elsewhere, which turned out to result in costing the system more than just doing the simple thing.)

The bigger issue for me is that this approach smells like premature optimization.

Donald Knuth sums my opinion about this up pretty well:

“The real problem is that programmers have spent far too much time worrying about efficiency in the wrong places and at the wrong times; premature optimization is the root of all evil (or at least most of it) in programming.”

As a a concrete example here, pass by value optimization of the Context structure meant that the Table of Contents in the context was not updated properly.  It also led to another set of duplicate to create the Table of Contents (so we actually re-ran the steps to generate them!) and additional code. 

Generally, I think it's better to design the code for correctness first, and then optimize the big areas first (algorithm changes ,etc.)  Details like this count as micro-optimization, and in the vast majority are simply not worth the effort.  (The exceptions would be for cases where the code paths are *extremely* hot and time sensitive, such as low level context switching code, code taken for every IP packet, or improvements to bulk cryptographic algorithms.)

However, *in this case*, it might make sense to just copy  the structure (pass by value).  It's big, which is why I didn't do it.  The desire to copy would be to ensure that have our own value so that if someone wanted to later change the template we wouldn't be impacted (because we don't reparse the templates).

Reply:
> So that's actually a fallacy, and IMO misguided.
> 
> Copy by value beats indirection in cases where you either are going to dereference every member anyway (so you're already copying anyway), or the object in question is small.
> 
> However, the total time difference is probably on the order of a few nanoseconds, and unless the dereferencing is done multiple times per member (it isn't here), it's likely a net wash (not even a few ns).
> 
Yes, that's my understanding as well. Also, the rule I've tried to follow so far was to avoid passing pointers if the argument does not need to be modified.


> (For the record, I've spent a significant part of my life building very high performance software for use in circumstances like high frequency trading, IP protocol routing, etc. At one point I was employed by Sun Microsystems as a contractor to optimize network performance to help them reach higher packet-per-second counts for IP forwarding. A significant part of the effort was often "undoing" attempts to optimize code paths that ultimately lead to more complexity elsewhere, which turned out to result in costing the system more than just doing the simple thing.)
> 

I believe you on this. I'm not sure there is more complexity elsewhere in the codebase here, but it's good to see that you have a lot of experience in this domain!

> The bigger issue for me is that this approach smells like premature optimization.
> 
> Donald Knuth sums my opinion about this up pretty well:
> 
> “The real problem is that programmers have spent far too much time worrying about efficiency in the wrong places and at the wrong times; premature optimization is the root of all evil (or at least most of it) in programming.”
> 

yes, I'm aware about this saying, and again, I wanted to follow to rule expressed above. My understanding is that the parser is by far the most time-consuming part of the whole process (by several orders of magnitude)

> As a a concrete example here, pass by value optimization of the Context structure meant that the Table of Contents in the context was not updated properly. It also led to another set of duplicate to create the Table of Contents (so we actually re-ran the steps to generate them!) and additional code.

I appreciate that you were able to find this bug and solve it :)

> 
> Generally, I think it's better to design the code for correctness first, and then optimize the big areas first (algorithm changes ,etc.) Details like this count as micro-optimization, and in the vast majority are simply not worth the effort. (The exceptions would be for cases where the code paths are _extremely_ hot and time sensitive, such as low level context switching code, code taken for every IP packet, or improvements to bulk cryptographic algorithms.)
> 
> However, _in this case_, it might make sense to just copy the structure (pass by value). It's big, which is why I didn't do it. The desire to copy would be to ensure that have our own value so that if someone wanted to later change the template we wouldn't be impacted (because we don't reparse the templates).

I'm sorry but this is where I'm lost: are you suggesting that in this case, passing by value is ok or not ok?

Reply:
In this case, it may be better to do it.  I don't think it's a big deal either way.  I think pass by value may be better here because it describes *semantics* more accurately -- that is we really mean that future changes to the structure won't have any effect.

In other words, you convinced, but not because of performance.  :-)

Reply:
I went ahead and made this change -- I'll commit it shortly.

Reply:
great, thanks!